x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/AuthorizationEngine.java

```diff
@@ -560,11 +573,20 @@ public IndexAuthorizationResult(boolean auditable, IndicesAccessControl indicesA
         }
 
         @Override
-        public String getFailureContext(RestrictedIndices restrictedIndices) {
+        public String getFailureContext(RequestInfo requestInfo, RestrictedIndices restrictedIndices) {
             if (isGranted()) {
                 return null;
             } else {
-                return getFailureDescription(indicesAccessControl.getDeniedIndices(), restrictedIndices);
+                assert indicesAccessControl != null;
+                String[] indices = RequestInfo.indices(requestInfo.getRequest());
+                if (indices == null) {
+                    return null;
+                }
+                Set<String> deniedIndices = Arrays.asList(indices)
+                    .stream()
+                    .filter(index -> false == indicesAccessControl.hasIndexPermissions(index))
+                    .collect(Collectors.toSet());
+                return getFailureDescription(deniedIndices, restrictedIndices);
```

Comment:
This new code is not exactly equivalent to the replaced one when there are datemaths.
But for a denied exception message, I think it is good enough.

Reply:
I also think it is OK and likely an improvement. The current code will report all concrete indices related to the unauthorized names. For example, if an alias (pointing to index) is unauthorized, the error message will say 
> ... unauthorized .. on indices [alias,index], ...

Strictly speaking, this is leaking information to unauthorized users. With this change, it will report only `alias`. It is also consistent with audit log which is a bonus. 

For datemath expression, I notice potentially a separate issue. The request `GET <index-{now/d}>/_search` returns `404` if the user does not have read privileges. But `GET index-2002.09.27/_search` returns `403` even though it is the same concrete index name as the first request. If we are going to handle datemath names as plain names, should they two behave the same way?

Assuming we fix the above, does it change how error message (and audit logs) work here? Right now, I cannot seem to trigger a 403 with datemath expression. If 404 is the only possibility with datemath, once we make it behave like concrete names, would it also show up as "resolved concrete name" in this error message and audit log?

Reply:
> For datemath expression, I notice potentially a separate issue. The request GET <index-{now/d}>/_search returns 404 if the user does not have read privileges. But GET index-2002.09.27/_search returns 403 even though it is the same concrete index name as the first request. If we are going to handle datemath names as plain names, should they two behave the same way?

Not on the TODO list, but it might get fixed as a collateral.

> Assuming we fix the above, does it change how error message (and audit logs) work here? Right now, I cannot seem to trigger a 403 with datemath expression. If 404 is the only possibility with datemath, once we make it behave like concrete names, would it also show up as "resolved concrete name" in this error message and audit log?

This is a very fine observation that this code here is executed for 403, but unauthorized datemaths generate 404 right now, so they do not triggering this, which is important because datemaths are not replaced in the request, so the error message would contain unevaluated datemaths....
As you can see it's sort of a mess when you get into the details, but it's usually not on the "important" code paths...
I just hope I nudge it into the right direction, step by step...

Reply:
> I just hope I nudge it into the right direction, step by step...

I mean that my objective in this PR is to remove "unauthorized" `IndicesAccessControl` from the access control map.
The unauthorized error message unfortunately needed a slight reworking, which I tried to keep to a minimum in order to not modify any more code now.